tracing/ops_tracing/backend.py

```diff
@@ -0,0 +1,111 @@
+# Copyright 2025 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
+# file except in compliance with the License. You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+"""The global implementation of the ops-tracing extension."""
+
+from __future__ import annotations
+
+import contextlib
+import logging
+from typing import TYPE_CHECKING, Generator
+
+from opentelemetry.sdk.resources import Resource
+from opentelemetry.sdk.trace import TracerProvider
+from opentelemetry.sdk.trace.export import BatchSpanProcessor
+from opentelemetry.trace import get_current_span, get_tracer_provider, set_tracer_provider
+
+if TYPE_CHECKING:
+    from ops.jujucontext import _JujuContext
+
+from .const import BUFFER_FILE, Config
+from .export import BufferingSpanExporter
+
+_exporter: BufferingSpanExporter | None = None
+
+
+class LogsToEvents(logging.Handler):
+    """An adaptor that convert log records to OTEL events."""
+
+    def emit(self, record: logging.LogRecord) -> None:
+        """Emit this log record as OTEL event."""
+        span = get_current_span()
+        if span and span.is_recording():
+            try:
+                message = record.getMessage()
+                level = record.levelname
+            except Exception as e:
+                # This should never happen, except if the charm includes custom logging
+                # library like structlog that enriches both the format and record attributes,
+                # or if the record format doesn't match the arguments.
+                message = f'log {record=} error {e}'
+                level = 'UNKNOWN'
+            span.add_event(message, {'level': level})
+
+
+@contextlib.contextmanager
+def setup(juju_context: _JujuContext, charm_class_name: str) -> Generator[None, None, None]:
+    """A context manager to control tracing lifespan."""
```

Comment:
This docstring doesn't really explain what it does. Also, why does `setup` need to be public? It's only called by `_main`, right?

Reply:
The function is not hosted in a private submodule:
`ops_tracing._backend.py: def setup`

And re-exported with a private name:
`ops_tracing.__init__.py: ... import setup as _setup`

The docstring has been updated.